File: scripts/instrumentation/optimized_mode_checker_code.py
import inspect
import json
import sys
import os
import io

# TRACE_FOLDER = ...
DEBUG = False
LOG_STEP = 10000 # total steps per log file
SKIP_STEP = 0 # skip the first n steps

total_step = -1
msg_queue = ""
call_stack = []
id2oid = {}
all_objects = []
global_object_idx = 0
first_occurrence_of_block = {}

### shims
shim_random_seq = []
shim_print_seq = []

def summary(obj, object_table):
    global global_object_idx
    typestr = type(obj).__name__
    
    ### If is primitive type, return directly.
    if obj == None:
        return None
    elif inspect.isbuiltin(obj):
        return ["U"]
    elif typestr in ["bool", "int", "float", "str"]:
        return obj
    
    if id(obj) not in id2oid:
        oid = "oid_" + str(global_object_idx)
        id2oid[id(obj)] = oid
        all_objects.append(obj) ### This will keep the refer count of the object >= 1. And make the object not be collected by GC.
        global_object_idx += 1
    else:
        oid = id2oid[id(obj)]
        if oid in object_table:
            return oid
        else:
            object_table[oid] = ["TO_BE_FILLED"]

    if (
        typestr == "list"
        or typestr == "tuple"
        # or isinstance(x, range)
        # or isinstance(x, zip)
        # or isinstance(x, enumerate)
        # or isinstance(x, deque)
        # or typestr == "<class 'collections.deque'>"
        or typestr == "dict_keys"
        or typestr == "dict_values"
        or typestr == "dict_items"
        # or typestr == "<class 'generator'>" # Lazy generator value
        # or isinstance(x, queue.Queue)
    ):
        val = []
        count = 0
        for v in obj:
            count += 1
            val.append(summary(v, object_table))
        
        object_table[oid] = ["L", count, val]
        return oid
    elif typestr == "set":
        val = []
        for v in obj:
            val.append(summary(v, object_table))
        
        object_table[oid] = ["SET", val]
        return oid
    elif typestr == "dict" or isinstance(obj, dict):
        val = []
        for k in obj:
            
            if type(k).__name__ in ["type", "function"]:
                key = "U"
                tmp = "U"
            else:
                tmp = summary(obj[k], object_table)
                if tmp == ["U"]:
                    continue
                key = summary(k, object_table)
            
            val.append([key, tmp])
        
        object_table[oid] = ["D", val]
        return oid
    elif typestr == "function":
        
        object_table[oid] = ["F", obj.__name__]
        return oid
    elif typestr == "type":
        
        object_table[oid] = ["C", obj.__name__]    
        return oid
    else:
        object_table[oid] = ["U"]
        return oid

def _instrument_random_shim(min, max):
    if total_step >= SKIP_STEP:
        shim_random_seq.append([min, max])
    return user_randint(min, max)

def _instrument_print_shim(*args, **kwargs):
    if total_step >= SKIP_STEP:
        output_capture = io.StringIO()
        sys.stdout = output_capture
        print(*args, **kwargs)
        IO_str = output_capture.getvalue()
        sys.stdout = sys.__stdout__
        shim_print_seq.append(IO_str)

    return print(*args, **kwargs)


def _instrument_begin(_call_info, _vars, _blockId):
    global total_step
    total_step += 1
    
    _from = call_stack[-1] if len(call_stack) > 0 else "START"
    call_stack.append(_blockId)
    _to = _blockId
    if total_step < SKIP_STEP:
        return
    
    vars = {}
    object_table = {}
    for var in _vars:
        if var.startswith("__"):
            continue
        vars[var] = summary(_vars[var], object_table)
    
    args = {}
    for arg in _call_info['args']:
        if arg.startswith("__"):
            continue
        args[arg] = summary(_call_info['args'][arg], object_table)
    call_info = {'args': args, 'func_name': _call_info['func_name']}
    _record_trace('begin', _blockId, _from, _to, vars, object_table, call_info)

def _instrument_throw(_err, _vars, _blockId):
    global total_step
    total_step += 1
    _from = _blockId
    call_stack.pop()
    _to = call_stack[-1] if len(call_stack) > 0 else "END"
    if total_step < SKIP_STEP:
        return
    
    vars = {}
    object_table = {}
    err = str(_err)
    for var in _vars:
        if var.startswith("__"):
            continue
        vars[var] = summary(_vars[var], object_table)
    _record_trace('throw', _blockId, _from, _to, vars, object_table, err)

def _instrument_return(_ret, _vars, _blockId):
    global total_step
    total_step += 1
    _from = _blockId
    call_stack.pop()
    _to = call_stack[-1] if len(call_stack) > 0 else "END"
    if total_step < SKIP_STEP:
        return _ret
    
    vars = {}
    object_table = {}
    ret = summary(_ret, object_table)
    for var in _vars:
        if var.startswith("__"):
            continue
        vars[var] = summary(_vars[var], object_table)
    _record_trace('return', _blockId, _from, _to, vars, object_table, ret)
    return _ret

def _record_trace(evt, blockId, _from, _to, vars, object_table, extra=None):
    global total_step, msg_queue
    if DEBUG and total_step >= 10 * LOG_STEP:
        exit()
    if total_step != SKIP_STEP and total_step % LOG_STEP == 0:
        filename = f"/_step_trace_from_{total_step - LOG_STEP}_to_{total_step}.log"
        print(f"[INFO] Writing to file {filename}")
        with open(TRACE_FOLDER + filename, 'w') as f:
            f.write(msg_queue)
        msg_queue = ""

    if _from not in first_occurrence_of_block: ### Record the first occurrence of the block.
        first_occurrence_of_block[_from] = total_step
        
    if evt == "begin":
        msg_queue += "Call|||From:" + str(_from) + "|||To:" + str(_to) + "|||" + json.dumps(vars)+ "|||" + json.dumps(object_table) + "|||" + json.dumps(extra)
    elif evt == "return":
        msg_queue += "Return|||From:" + str(_from) + "|||To:" + str(_to) + "|||" + json.dumps(vars)+ "|||" + json.dumps(object_table) + "|||" + json.dumps(extra)
    elif evt == "throw":
        msg_queue += "Throw|||From:" + str(_from) + "|||To:" + str(_to) + "|||" + json.dumps(vars)+ "|||" + json.dumps(object_table) + "|||" + json.dumps(extra)
    else:
        ...
    
    if shim_random_seq != []:
        msg_queue += "|||" + json.dumps(shim_random_seq)
        shim_random_seq.clear()
    else:
        msg_queue += "|||[]"
    
    if shim_print_seq != []:
        msg_queue += "|||" + json.dumps(shim_print_seq)
        shim_print_seq.clear()
    else:
        msg_queue += "|||[]"
    
    msg_queue += "\n"
    
    if evt in ["throw", "return"] and blockId == 0: ### End of the program
        filename = f"/_step_trace_from_{total_step - total_step % LOG_STEP}_to_{total_step - total_step % LOG_STEP + LOG_STEP}.log" 
        with open(TRACE_FOLDER + filename, 'w') as f:
            f.write(msg_queue)
        with open(TRACE_FOLDER + '/_step_trace_info.json', 'w') as f:
            trace_info = {'total_step': total_step, 'log_step': LOG_STEP, 'skip_step': SKIP_STEP}
            f.write(json.dumps(trace_info, indent=4))
        with open(TRACE_FOLDER + '/_first_occurrence_of_block.json', 'w') as f:
            f.write(json.dumps(first_occurrence_of_block, indent=4))
        msg_queue = ""
        

File: scripts/instrumentation/optimized_mode_checker_code.js
// const TRACE_FOLDER = ...
const DEBUG = false
const LOG_STEP = 10000 // total steps per log file
const SKIP_STEP = 0 // skip the first n steps

let total_step = -1
let call_stack = []
let msg_queue = []
let object2oid = new Map();
let global_object_idx = 0

let shim_random_seq = []
let shim_print_seq = []

function inTimeCompare(x, x_object_table, y, verified_object_table){
    let tpx_a = typeof x;
    // console.log(x, tpx_a)
    if (x === "undefined"){
        return [true, null];
    }

    if (tpx_a === "string" && !x.startsWith("oid_")){
        if (Number(x) === y || y !== undefined && y.toString() === x || x === y) return [true, null];
        return [false, null];
    }
    
    if (tpx_a === "number") {
        if (y === null) return [false, null];
        if (Math.abs(x - y) < 0.001) return [true, null];
        else return [false, null];
    }
        
    if(tpx_a === "boolean" || x === null) {
        return [x === y, null];
    }

    if (x[0] == "U"){
        return [true, null];
    }
    else if (y === null){ /// This is important. Since later we may access "y.constructor" and "y.name", we need to make sure y is not null
        return [false, null];
    }

    if (tpx_a === "string" && x.startsWith("oid_")){
        if (!object2oid.has(y)){
            var oid = "oid_" + (global_object_idx).toString();
            object2oid.set(y, oid);
            global_object_idx += 1
            if (oid !== x){
                // console.log("Object ID Mismatch", x, oid)
                return [false, "oid mismatch"];
            }
        }
        else{
            var oid = object2oid.get(y);
            if (verified_object_table.hasOwnProperty(oid)){
                return [true, null];
            }
            else{
                if (oid !== x){
                    // console.log("Object ID Mismatch", x, oid)
                    return [false, "oid mismatch"];
                }
                verified_object_table[oid] = true;
            }
        }
    
        var x_obj = x_object_table[x]; 
        if (x_obj[0] === "U" && y === "undefined") return [false, null]; /// Consider as error.
        if (y === undefined) return [false, null];

        if (x_obj[0] === "F") {
            if (x_obj[1] === "<lambda>") {
                return [true, null];
            }
            return [x_obj[1] === y.name, null];
        }   
        else if (x_obj[0] === "L") {
            if (y.constructor === Array && x_obj[1] === y.length) {
                for (let i = 0; i < x_obj[1]; i++) {
                    var [eq, rs] = inTimeCompare(x_obj[2][i], x_object_table, y[i], verified_object_table)
                    if (!eq) {
                        return [false, null];
                    }
                }
                return [true, null];
            }
            else{
                return [false, null];
            }
        }
        else if (x_obj[0] === "SET") {
            if (y.constructor === Set && x_obj[1].length === y.size) {
                for (let v of x_obj[1]) {
                    if (!y.has(v)) {
                        return [false, null];
                    }
                }
                return [true, null];
            }
            else{
                return [false, null];
            }
        }
        else if (x_obj[0] === "D") {
            if (y.constructor === Object) {
                for (let k of x_obj[1]) {
                    if (!(typeof k[0] === "string" && k[0].startsWith("oid_")) && k[0] !== "U"){ /// Complex type should not be the dict key 
                        var [eq, rs] = inTimeCompare(k[1], x_object_table, y[k[0]], verified_object_table);
                    }
                    else{
                        var eq = true
                    }
                    if (!eq) {
                        return [false, null];
                    }
                }
                return [true, null];
            }
            else{
                return [false, null];
            }
        }
        else{
            return [true, null];
        }
    }
    return [true, null];
}

function _instrument_random_shim(min, max) {
    if (total_step >= SKIP_STEP){
        shim_random_seq.push([min, max])
    }
    return user_randint(min, max)
}

function _instrument_print_shim(...args) {
    if (total_step >= SKIP_STEP){
        var capturedOutput = args.map(arg => String(arg)).join(" ") + "\n";
        shim_print_seq.push(capturedOutput)
    }
    return console.log(...args)
}

function _instrument_begin(_vars, blockId) {
    total_step += 1
    var _from = call_stack.length > 0 ? call_stack[call_stack.length - 1] : "START";
    call_stack.push(blockId);
    var _to = blockId;
    if (total_step < SKIP_STEP){
        return
    }
    _compare_trace("Call", blockId, _from, _to, _vars, null)
}

function _instrument_throw(_err, _vars, blockId) {
    total_step += 1
    var _from = blockId;
    call_stack.pop();
    var _to = call_stack.length > 0 ? call_stack[call_stack.length - 1] : "END";
    if (total_step < SKIP_STEP){
        return
    }
    _compare_trace("Throw", blockId, _from, _to, _vars, _err.message)
}

function _instrument_return(_ret, _vars, blockId) {
    total_step += 1
    var _from = blockId;
    call_stack.pop();
    var _to = call_stack.length > 0 ? call_stack[call_stack.length - 1] : "END";
    if (total_step < SKIP_STEP){
        return _ret
    }
    _compare_trace("Return", blockId, _from, _to, _vars, _ret)
    return _ret;
}


function loadMsgQueue(filename) {
    /// load message que from the file
    const fs = require('fs');
    const path = require('path');
    // const filePath = path.join(__dirname, filename);
    const filePath = filename; // use relative path
    if (fs.existsSync(filePath)) {
        let data = fs
            .readFileSync(filePath)
            .toString();
        let messages = data.split("\n");
        for (let i = 0; i < messages.length; i++){
            let msg_sections = messages[i].split("|||");
            if (msg_sections.length < 4) {
                continue;
            }
            let new_msg = {}
            new_msg["evt"] = msg_sections[0];
            new_msg["from"] = msg_sections[1].split(":")[1]
            new_msg["to"] = msg_sections[2].split(":")[1]
            try{
                new_msg["vars"] = JSON.parse(msg_sections[3])
            }
            catch(e){
                if (msg_sections[3].includes("Infinity")){
                    new_msg["vars"] = JSON.parse(msg_sections[3].replace(/Infinity/g, '"Infinity"'))
                }
                else console.log(msg_sections[3])
            }
            
            try{
                new_msg["object_table"] = JSON.parse(msg_sections[4])
            }
            catch(e){
                if (msg_sections[4].includes("Infinity")){
                    new_msg["object_table"] = JSON.parse(msg_sections[4].replace(/Infinity/g, '"Infinity"'))
                }
                else console.log(msg_sections[4])
            }
            
            
            if (new_msg["evt"] !== "Call"){
                try{
                    new_msg["extra"] = JSON.parse(msg_sections[5])
                }
                catch(e){
                    if (msg_sections[5].includes("Infinity")){
                        new_msg["extra"] = JSON.parse(msg_sections[5].replace(/Infinity/g, '"Infinity"'))
                    }
                    else console.log(msg_sections[5])
                }
            }
            
            new_msg["random_shim"] = JSON.parse(msg_sections[6])
            new_msg["print_shim"] = JSON.parse(msg_sections[7])

            msg_queue.push(new_msg)
        }
    }
    else{
        throw new Error("File not found")
    }
    return;
}

function _compare_trace(evt, blockId, _from, _to, vars, extra){
    try{
        if (total_step % LOG_STEP === 0){
            let filename = TRACE_FOLDER + `/_step_trace_from_${total_step}_to_${total_step + LOG_STEP}.log`;
            msg_queue = [];
            console.log("[INFO] Checking Step:", [total_step, total_step + LOG_STEP])
            console.log("[INFO] Loading Trace Log File: ", filename)
            loadMsgQueue(filename)
        }

        if (DEBUG && total_step >= 10 * LOG_STEP){
            throw new Error("STOP EXEC")
        }
        
        var extra_str = ""; // circular structure
        try {
            extra_str = JSON.stringify(extra)
        }
        catch(e){
            extra_str = "complex value that cannot be JSON.stringified"
        }




        var msg = msg_queue[total_step % LOG_STEP];

        /// Check the shim functions
        if (msg['random_shim'].length === shim_random_seq.length){
            for (let i = 0; i < msg['random_shim'].length; i++){
                if (msg['random_shim'][i][0] !== shim_random_seq[i][0] || msg['random_shim'][i][1] !== shim_random_seq[i][1]){
                    let err_msg = {
                        "Step": total_step,
                        "Reason": "Random Shim Call Value Mismatch",
                        "Expected": msg['random_shim'][i],
                        "Got": shim_random_seq[i]
                    }
                    throw new Error(JSON.stringify(err_msg))
                }
            }
        }
        else{
            let err_msg = {
                "Step": total_step,
                "Reason": "Random Shim Call Number Mismatch",
                "Expected": msg['random_shim'].length,
                "Got": shim_random_seq.length
            }
            throw new Error(JSON.stringify(err_msg))
        }

        shim_random_seq = []
        
        if (msg['print_shim'].length === shim_print_seq.length){
            for (let i = 0; i < msg['print_shim'].length; i++){
                if (msg['print_shim'][i] !== shim_print_seq[i]){
                    let err_msg = {
                        "Step": total_step,
                        "Reason": "Print Shim Call Value Mismatch",
                        "Expected": msg['print_shim'][i],
                        "Got": shim_print_seq[i]
                    }
                    throw new Error(JSON.stringify(err_msg))
                }
            }
        }
        else{
            let err_msg = {
                "Step": total_step,
                "Reason": "Print Shim Call Number Mismatch",
                "Expected": msg['print_shim'].length,
                "Got": shim_print_seq.length
            }
            throw new Error(JSON.stringify(err_msg))
        }

        shim_print_seq = []

        if (evt !== msg['evt']){
            let err_msg = {
                "Step": total_step,
                "Reason": "Event Mismatch",
                "ExpectedEvt": msg['evt'],
                "ExpectedFrom": msg['from'],
                "ExpectedTo": msg['to'],
                "ExpectedExtra": msg['extra'],
                "ExpectedObjectTable": msg['object_table'],
                "GotEvt": evt,
                "GotFrom": _from,
                "GotTo": _to,
                "GotExtra": extra_str
            }
            throw new Error(JSON.stringify(err_msg))
        }

        if (msg['from'] !== _from.toString() || msg['to'] !== _to.toString()){
            let err_msg = {
                "Step": total_step,
                "Reason": "Jump Target Mismatch",
                "ExpectedEvt": msg['evt'],
                "ExpectedFrom": msg['from'],
                "ExpectedTo": msg['to'],
                "ExpectedExtra": msg['extra'],
                "ExpectedObjectTable": msg['object_table'],
                "GotEvt": evt,
                "GotFrom": _from,
                "GotTo": _to,
                "GotExtra": extra_str
            }
            throw new Error(JSON.stringify(err_msg))
        }

        /// First check return value, then check value of vairables.
        /// The order of building the oid table should be same as python side.
        if (evt === "Call"){
            // return
        }
        else if (evt === "Return"){
            var [eq, rs] = inTimeCompare(msg['extra'], msg['object_table'], extra, {});
            if (!eq){
                let err_msg = {
                    "Step": total_step,
                    "Reason": "Return Value Mismatch",
                    "ExpectedEvt": msg['evt'],
                    "ExpectedFrom": msg['from'],
                    "ExpectedTo": msg['to'],
                    "ExpectedExtra": msg['extra'],
                    "ExpectedObjectTable": msg['object_table'],
                    "GotEvt": evt,
                    "GotFrom": _from,
                    "GotTo": _to,
                    "GotExtra": extra_str,
                    "ReturnMismatchReason": rs
                }
                // console.log(JSON.stringify(msg, null, 4))
                throw new Error(JSON.stringify(err_msg))
            }
            // return
        }
        else if (evt ==="Throw"){
            // return
        }
        else{
            throw new Error("Evt not supported")
        }


        for (let varName in msg['vars']) {
            var [eq, rs] = inTimeCompare(msg['vars'][varName], msg['object_table'], vars[varName], {});
            if (!eq){
                var value_str = "";
                try {
                    value_str = JSON.stringify(vars[varName])
                }
                catch(e){
                    value_str = "complex value that cannot be JSON.stringified"
                }

                let err_msg = {
                    "Step": total_step,
                    "Reason": "Variable Mismatch",
                    "ExpectedEvt": msg['evt'],
                    "ExpectedFrom": msg['from'],
                    "ExpectedTo": msg['to'],
                    "ExpectedExtra": msg['extra'],
                    "ExpectedObjectTable": msg['object_table'],
                    "GotEvt": evt,
                    "GotFrom": _from,
                    "GotTo": _to,
                    "GotExtra": extra_str,
                    "VarName": varName,
                    "ExpectedValue": msg['vars'][varName],
                    "GotValue": value_str,
                    "VarMismatchReason": rs
                }
                throw new Error(JSON.stringify(err_msg))
            }
        }


        if (["Throw", "Return"].includes(evt) && blockId == 0){ /// End of the program
            let filename = TRACE_FOLDER + `/_step_check_result.json`;
            let msg = {
                "step": total_step,
            }
            const fs = require('fs');
            fs.writeFileSync(filename, JSON.stringify(msg, null, 4));
            // process.exit(0)
        }
    }
    catch(e){
        /// Early end the tracing
        /// Save in Json
        let filename = TRACE_FOLDER + `/_step_check_result.json`;
        let msg = {
            "step": total_step,
            "err": JSON.parse(e.message),
            "call_stack": call_stack,
        }
        const fs = require('fs');
        fs.writeFileSync(filename, JSON.stringify(msg, null, 4));
        process.exit(0)
    }
}

